### PR TITLE
SoundWire/ASoC-hda-mlink: fixes

### PR DESCRIPTION
**Pull request by plbossart**
Add helpers and fix previous functionality that didn't quite work.

### CODE REVIEW

File: sound/soc/sof/intel/hda-mlink.c [hunk 1]
```diff
@@ -340,6 +340,21 @@ static void hdaml_link_set_lsdiid(u32 __iomem *lsdiid, int dev_num)
 	writel(val, lsdiid);
 }
 
+static void hdaml_shim_map_stream_ch(u16 __iomem *pcmsycm, int lchan, int hchan,
+				     int stream_id, int dir)
+{
+	u16 val;
+
+	val = readw(pcmsycm);
```
Comment (ranj063) [inline sound/soc/sof/intel/hda-mlink.c:349]: minor nit. can be moved to the declaration line

**Review by ranj063 — DISMISSED** [review]
looks good except for a minor nit

File: sound/soc/sof/intel/hda-mlink.c [hunk 2]
```diff
@@ -96,7 +96,7 @@ struct hdac_ext2_link {
  */
 
 static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
-			  void __iomem *ml_addr, int link_idx)
+			  void __iomem *remap_addr, void __iomem *ml_addr, int link_idx)
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-mlink.c:100]: Why do you need to pass a separate `ml_addr`? The function uses `ml_addr` or `hlink->ml_addr` interchangeably.

File: sound/soc/sof/intel/hda-mlink.c [hunk 3]
```diff
@@ -140,6 +140,7 @@ static int hdaml_lnk_enum(struct device *dev, struct hdac_ext2_link *h2link,
 
 	switch (h2link->elid) {
 	case AZX_REG_ML_LEPTR_ID_SDW:
+		h2link->instance_offset = AZX_REG_SDW_INSTANCE_OFFSET;
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-mlink.c:144]: is this a fixup! or a Fixes tag needed?

File: sound/soc/sof/intel/hda-mlink.c [hunk 4]
```diff
@@ -756,6 +771,35 @@ int hdac_bus_eml_sdw_set_lsdiid(struct hdac_bus *bus, int sublink, int dev_num)
...
+	struct hdac_ext2_link *h2link;
+	u16 __iomem *pcmsycm;
+	u16 val;
+
+	h2link = find_ext2_link(bus, true, AZX_REG_ML_LEPTR_ID_SDW);
+	if (!h2link)
+		return -ENODEV;
+
+	pcmsycm = h2link->base_ptr + h2link->shim_offset +
+		h2link->instance_offset * sublink + 0x16 + 0x4 * y;
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-mlink.c]: `0x16` and `0x4` might needs some explanation or define?

File: sound/soc/sof/intel/hda-mlink.c [hunk 5]
```diff
@@ -756,6 +771,35 @@ int hdac_bus_eml_sdw_set_lsdiid(struct hdac_bus *bus, int sublink, int dev_num)
 	return 0;
 } EXPORT_SYMBOL_NS(hdac_bus_eml_sdw_set_lsdiid, SND_SOC_SOF_HDA_MLINK);
 
+int hdac_bus_eml_sdw_map_stream_ch(struct hdac_bus *bus, int sublink, int y,
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-mlink.c:779]: what is the `y` parameter?

File: sound/soc/sof/intel/hda-mlink.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:349]: indeed, but the style for this file is to keep variables and first read separate.

File: sound/soc/sof/intel/hda-mlink.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:100]: the confusion between ml_addr and hlink->ml_addr is something we could add in a separate patch. We still need the ml_addr to figure out where the link descriptors are stored.

File: sound/soc/sof/intel/hda-mlink.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:144]: it was not used so far so there's no real need for any stable backport based on a Fixes: tag

File: sound/soc/sof/intel/hda-mlink.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:779]: it's named after the register: PCMSyCM.

y is the PDI number in this case.

I could add comments to make it self-explanatory.

File: sound/soc/sof/intel/hda-mlink.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c]: yes, but I wasn't really able to find any names. This comes straight from the HAS, as is.

File: sound/soc/sof/intel/hda-mlink.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:100]: added a patch

File: sound/soc/sof/intel/hda-mlink.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:144]: added comment in commit message

File: sound/soc/sof/intel/hda-mlink.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:779]: added comments and macro

File: sound/soc/sof/intel/hda-mlink.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c]: added macro

**Review by ranj063 — DISMISSED** [review]

**Comment by plbossart** [comment]
@bardliao @ujfalusi we need this in the topic/sof-dev and topic/sof-dev-rebase quickly before we start upstreaming for the next cycle - probably next week if the merge window closes as expected this Sunday

File: sound/soc/sof/intel/hda-mlink.c [hunk 6]
```diff
@@ -756,6 +772,40 @@ int hdac_bus_eml_sdw_set_lsdiid(struct hdac_bus *bus, int sublink, int dev_num)
...
+	mutex_lock(&h2link->eml_lock);
+
+	hdaml_shim_map_stream_ch(pcmsycm, 0, hweight32(channel_mask),
+				 stream_id, dir);
+
+	mutex_unlock(&h2link->eml_lock);
+
+	val = readw(pcmsycm);
+
+	dev_dbg(bus->dev, "channel_mask %d stream_id %d dir %d pcmscm %#x\n",
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-mlink.c]: dump channel_mask in hex number ?

**Review by RanderWang — DISMISSED** [review]

File: sound/soc/sof/intel/hda-mlink.c [hunk 7]
```diff
@@ -756,6 +772,40 @@ int hdac_bus_eml_sdw_set_lsdiid(struct hdac_bus *bus, int sublink, int dev_num)
 	return 0;
 } EXPORT_SYMBOL_NS(hdac_bus_eml_sdw_set_lsdiid, SND_SOC_SOF_HDA_MLINK);
 
+/*
+ * the 'y' parameter comes from the PCMSyCM hardware register naming. 'y' refers to the
+ * PDI index, i.e. the FIFO used for RX or TX
+ */
+int hdac_bus_eml_sdw_map_stream_ch(struct hdac_bus *bus, int sublink, int y,
+				   int channel_mask, int stream_id, int dir)
```
Comment (bardliao) [inline sound/soc/sof/intel/hda-mlink.c:780]: Should `channel_mask` be unsigned?

Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c:780]: i'll keep it as int, we are not using unsigned int anywhere

File: sound/soc/sof/intel/hda-mlink.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/intel/hda-mlink.c]: done, thanks for the suggestion

**Review by bardliao — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Pull request merged**